### PR TITLE
List filtering

### PR DESCRIPTION
**Pull request by tkuhlman**
POC of how list filtering could be added

### CODE REVIEW

**Comment by tkuhlman** [comment]
This is no longe a POC I completed the code and it is ready for review

File: gql/comparator.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+	"sync"
+)
+
+// A Comparator checks whether a value Matches.
+// Each NewListOperation returns a comparator to be used in List filtering.
+type Comparator interface {
+	Match(interface{}) bool
+}
+
+// NotCompartor simply does a logical not on the Match result of its child.
```
Comment (gannett-ggreer) [inline gql/comparator.go]:
```suggestion
// NotComparator simply does a logical not on the Match result of its child.
```

File: gql/builder.go [hunk 2]
```diff
@@ -325,6 +342,58 @@ func (ob *ObjectBuilder) resolveObjectByName(p graphql.ResolveTypeParams) *graph
...
+func ResolveListField(name string, parent string) graphql.FieldResolveFn {
+	return func(p graphql.ResolveParams) (interface{}, error) {
+		filter, err := newListFilter(p.Args[filterArgumentName])
+		if err != nil {
+			return nil, err
+		}
+
+		resolve := ResolveByField(name, parent)
+
+		fullList, err := resolve(p)
```
Comment (enigmatic00) [inline gql/builder.go]: nit: we don't know that this is a list yet, so can we not call this fullList?

File: gql/comparator.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+		values[i] = item
+	}
+
+	return inComparator{values: values}, nil
+}
+
+// NewIntegerComparator returns a NewListOperator for integer operations supported by integerComparator,
+// specifically <, <=, > and >=.
+// The returned NewListOperation expects an int with the key "Value" as part of the given argument.
+func NewIntegerComparator(op string) NewListOperation {
```
Comment (gannett-ggreer) [inline gql/comparator.go]: Maybe I'll find the reason as I continue reading, but this one is an odd duck compared to the others in that it returns `NewListOperation` instead of `Comparator`, which makes its error handling really weird.

File: gql/comparator.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+	eq, err := NewEqualComparator(arg)
+	if err != nil {
+		return nil, err
+	}
+	return NotComparator{Child: eq}, nil
+}
+
+// NewInComparator returns a comparator which returns true if the operand matches any of the given values.
+// The comparator expects a list of strings under the key "Values" as part of the given argument.
+func NewInComparator(arg map[string]interface{}) (Comparator, error) {
```
Comment (enigmatic00) [inline gql/comparator.go]: NewIntComparator?

Comment (enigmatic00) [reply gql/comparator.go]: nope, inComparator is fine

File: gql/comparator.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+	if value.Kind() != reflect.Slice {
+		return nil, errors.New("filter argument 'Values' should be a list of strings")
+	}
+
+	values := make([]string, value.Len())
+	for i := 0; i < value.Len(); i++ {
+		rawItem := value.Index(i).Interface()
+		item, ok := rawItem.(string)
+		if !ok {
+			return nil, errors.New("filter argument 'Values' should be a list of strings")
```
Comment (enigmatic00) [inline gql/comparator.go]: what about a list of integers?

File: gql/comparator_test.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+			operation:   NewEqualComparator,
+			arguments:   map[string]interface{}{"Value": "a"},
+			operand:     3,
+			want:        false,
+		},
+		{
+			description:          "NewEqualComparator invalid argument, nil",
+			operation:            NewEqualComparator,
+			arguments:            nil,
+			wantInvalidOperation: true,
```
Comment (enigmatic00) [inline gql/comparator_test.go]: would "" be treated as an invalidOperation or just return a false?

File: gql/comparator_test.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+		},
+		{
+			description: "int equal valid and true",
+			operation:   NewEqualComparator,
+			arguments:   map[string]interface{}{"Value": 1},
+			operand:     1,
+			want:        true,
+		},
+		{
+			description: "int equal valid and false",
```
Comment (enigmatic00) [inline gql/comparator_test.go]: there's no reason for these to not work , but to be thorough, can you add test cases for 0 & a negative number too?

File: gql/comparator_test.go [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+			wantInvalidOperation: true,
+		},
+		{
+			description:          "NOT IN invalid argument, string values",
+			operation:            NewNotInComparator,
+			arguments:            map[string]interface{}{"Values": "a, b, c"},
+			wantInvalidOperation: true,
+		},
+		{
+			description:          "NOT IN invalid argument, integer in lis",
```
Comment (enigmatic00) [inline gql/comparator_test.go]: integer in lis*t*

File: gql/filter.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+		"==":          NewEqualComparator,
+		"!=":          NewNotEqualComparator,
+		">":           NewIntegerComparator(">"),
+		">=":          NewIntegerComparator(">="),
+		"<":           NewIntegerComparator("<"),
+		"<=":          NewIntegerComparator("<="),
+		"limitLength": NewLimitLengthComparator,
+		"IN":          NewInComparator,
+		"NOT IN":      NewNotInComparator,
+		"in":          NewInComparator,
```
Comment (enigmatic00) [inline gql/filter.go]: we should be consistent and only support 1 case for the operations. IMO, they should be `LIMIT, IN, NOT IN`

File: gql/comparator.go [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+	case ">":
+		return in > c.value
+	case ">=":
+		return in >= c.value
+	case "<":
+		return in < c.value
+	case "<=":
+		return in <= c.value
+	default:
+		return false
```
Comment (gannett-ggreer) [inline gql/comparator.go]: Seems more appropriate to `panic` here since this should never happen.

File: gql/comparator.go (hunk 3 above)
Comment (tkuhlman) [reply gql/comparator.go]: Yeah it is different but saves on a lot of repeated code.

File: gql/comparator.go (hunk 4 above)
Comment (tkuhlman) [reply gql/comparator.go]: Nope, IN like the SQL operation

File: gql/builder.go (hunk 2 above)
Comment (tkuhlman) [reply gql/builder.go]: fixing

File: gql/comparator.go (hunk 5 above)
Comment (tkuhlman) [reply gql/comparator.go]: We could add that support but at least in the initial MVP there was no use case for it

File: gql/comparator.go (hunk 10 above)
Comment (tkuhlman) [reply gql/comparator.go]: Yeah it should be impossible, I'll update

File: gql/comparator.go [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+	}
+}
+
+type limitLength struct {
+	limit int
+	count int
+	mtx   sync.Mutex
```
Comment (gannett-ggreer) [inline gql/comparator.go]: Since `ResolveListField` does the matches sequentially, I wouldn't think you would need to have a mutex here.  If you *do* need a mutex here, how confident are you the `count` is accurate and not being shared between unrelated filter lists?

File: gql/comparator_test.go (hunk 6 above)
Comment (tkuhlman) [reply gql/comparator_test.go]: the arguments needs to be a `map[string]interface{}` so "" would be a compile error

File: gql/comparator_test.go (hunk 7 above)
Comment (tkuhlman) [reply gql/comparator_test.go]: sure

File: gql/filter.go (hunk 9 above)
Comment (tkuhlman) [reply gql/filter.go]: I admit I don't like limitLength now that I see it next to the others. I could go either way with multiple case, so unless there are other opinions I'll stick to these recommendations.

File: gql/comparator_test.go (hunk 6 above)
Comment (enigmatic00) [reply gql/comparator_test.go]: I meant something like `map[string]interface{}{"Value": ""},`

File: gql/comparator.go (hunk 11 above)
Comment (tkuhlman) [reply gql/comparator.go]: The parallel handling of queries in the GraphQL library we are currently using is a bit of a mess, however in general when it hits a list the resolving of each list item is done in its own go routine. That allows for parallel retrieval in the case of the resolving of a field requiring a DB call. Basically that means I do think I need a mutex here.

Regarding how confident I am it is not being shared between unrelated filter lists I'm confident. The comparator is made as part of the filter which is created in ResolveListField which runs fresh for each query being processed and multiple times depending on which field it is associated with.

File: gql/filter.go [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+	// The default implementations are pre-populated and additional operations can be added before running the
+	// ObjectBuilder.
+	ListOperations = map[string]NewListOperation{
+		"==":     NewEqualComparator,
+		"!=":     NewNotEqualComparator,
+		">":      NewIntegerComparator(">"),
+		">=":     NewIntegerComparator(">="),
+		"<":      NewIntegerComparator("<"),
+		"<=":     NewIntegerComparator("<="),
+		"LIMIT":  NewLimitLengthComparator,
```
Comment (enigmatic00) [inline gql/filter.go]: 💯

File: gql/comparator_test.go (hunk 6 above)
Comment (tkuhlman) [reply gql/comparator_test.go]: Oh that would be valid, a comparison to the empty string. I'll add in a case for it.

File: gql/filter.go (hunk 9 above)
Comment (gannett-ggreer) [reply gql/filter.go]: Why not case insensitive?  SQL is.   And I agree with Devo about `limit` rather than `limitLength`.

Comment (tkuhlman) [reply gql/filter.go]: I could change it to make all operations case insensitive, though in retrospect probably best to do this for all operations rather than just selectively if we decide that is best.

File: gql/comparator.go (hunk 11 above)
Comment (gannett-ggreer) [reply gql/comparator.go]: Given the place where you have inserted this resolve function, shouldn't all the items in that list have already been finished by the time you're doing the match processing? So the limit itself wouldn't need to be parallel because all the parallel bits happen lower and are done?

Comment (tkuhlman) [reply gql/comparator.go]: hmm, maybe that is right the parallel bits have already been done. Though that sounds right that code is complicated and this is usable by other uses cases since NewLimitLengthComparator is public. That makes me inclined to keep the mutex for the safety it brings.

File: gql/comparator.go (hunk 3 above)
Comment (tkuhlman) [reply gql/comparator.go]: good point, fixing

File: gql/comparator.go (hunk 11 above)
Comment (gannett-ggreer) [reply gql/comparator.go]: Sounds reasonable to me.

I worry, however, that anything that does take advantage of the mutex will also have to deal with a non-deterministic return value.
